src/hooks/UseForm/useForm/useForm.ts

```diff
@@ -0,0 +1,458 @@
+    // @ts-expect-error TDimensions defaults to 0 if dimensions is not defined, so this is okay
+  }: UseFormOptions<TDimensions> = { dimensions: 0 },
+): UseFormReturn<TData, TDimensions> => {
+  const fieldsGraph = useRef<FormData<Fields<TData, TDimensions>, TDimensions>>(
+    new Graph<Fields<TData, TDimensions>, TDimensions>({ dimensions }),
+  )
+  const [
+    errors,
+    { updateVertex: updateErrorsVertex, set: setErrors, pruneVertex: pruneErrorsVertex },
+  ] = useGraph<Errors<TData>, TDimensions>({ dimensions })
+  const [
+    touched,
+    {
+      someVertex: touchedAtSomeVertex,
+      updateVertex: updateTouchedVertex,
+      pruneVertex: pruneTouchedVertex,
+    },
+  ] = useGraph<Touched<TData>, TDimensions>({ dimensions })
+  const [
+    changed,
+    {
+      someVertex: changedAtSomeVertex,
+      updateVertex: updateChangedVertex,
+      updateAllVertices: updateAllChangedVertices,
+      pruneVertex: pruneChangedVertex,
+    },
+  ] = useGraph<Changed<TData>, TDimensions>({ dimensions })
+  const hasBegun = useMemo(
+    () =>
+      !!touched &&
+      touchedAtSomeVertex(fields =>
+        fields ? Object.values(fields).some(fieldHasBeenTouched => fieldHasBeenTouched) : false,
+      ),
+    [touched, touchedAtSomeVertex],
+  )
+  const hasChangedWithoutSubmit = useMemo(
+    () =>
+      !!changed &&
+      changedAtSomeVertex(fields =>
+        fields ? Object.values(fields).some(fieldHasBeenChanged => fieldHasBeenChanged) : false,
+      ),
+    [changed, changedAtSomeVertex],
+  )
+
+  const updateFieldsVertex = useCallback(
+    (
+      fieldName: keyof TData,
+      updater: (
+        current?: Fields<TData, TDimensions>[keyof TData],
+      ) => Fields<TData, TDimensions>[keyof TData],
+      coordinates?: CoordinatesOrNever<TDimensions, CoordinatesOfLength<TDimensions>>,
+    ) =>
+      fieldsGraph.current.updateVertex(
+        fields =>
+          fields
+            ? { ...fields, [fieldName]: updater(fields?.[fieldName]) }
+            : ({ [fieldName]: updater() } as Fields<TData, TDimensions>),
+        coordinates,
+      ),
+    [],
+  )
+
+  const setFieldsVertex = useCallback(
+    <TFieldName extends keyof TData>(
+      fieldName: TFieldName,
+      newValue: Fields<TData, TDimensions>[TFieldName],
+      coordinates?: CoordinatesOrNever<TDimensions, CoordinatesOfLength<TDimensions>>,
+    ) =>
+      fieldsGraph.current.updateVertex(
+        fields =>
+          fields
+            ? { ...fields, [fieldName]: newValue }
+            : ({ [fieldName]: newValue } as Fields<TData, TDimensions>),
+        coordinates,
+      ),
+    [],
+  )
+
+  const touch = useCallback(
+    (
+      fieldName: keyof TData,
+      coordinates?: CoordinatesOrNever<TDimensions, CoordinatesOfLength<TDimensions>>,
+    ) => {
+      updateFieldsVertex(fieldName, fields => ({ ...fields!, hasBeenTouched: true }), coordinates)
+      updateTouchedVertex(fields => ({ ...fields!, [fieldName]: true }), coordinates)
```

Comment:
Because these functions are only called by events fired on a registered field, it is guaranteed that those fields will already exist here. That said, I updated them to check anyway in case the functions are ever used elsewhere.